### PR TITLE
Color Configs

### LINKED ISSUE #9: Allow people to change the colors by using a specific section in the config file
The way this should be done is by, first, abstracting the “friendly” and al. functions (which basically wrap strings with `Colored` methods). That abstraction should be using the configuration file to:

1. Support terminal colors, like `color0`, `color8`, etc.
2. RGB colors, via hexadecimal strings, like `#30ac45`.
3. Support styles (italics, bright, dimmed, underline, strikethrough).

### PR DESCRIPTION
**Pull request by LordMZTE**
this should close #9

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -23,5 +23,7 @@ env_logger = "0.8.1"
 log = "0.4.11"
 serde = { version = "1", features = ["derive"] }
 serde_json = "1"
+serde_test = "1"
```
Comment (phaazon) [inline Cargo.toml:26]: This should go in a dedicated section for build only.

File: src/cli.rs [hunk 2]
```diff
@@ -135,11 +135,8 @@ pub fn list_tasks(
     display_task_header(config, &display_opts);
   }
 
-  let mut parity = true;
```
Comment (phaazon) [inline src/cli.rs:138]: Why did you remove this code? This is a visual regression.

File: src/cli.rs [hunk 3]
```diff
@@ -176,7 +173,7 @@ pub fn add_task(
   let display_opts = DisplayOptions::new(config, once((uid, &task)));
 
   display_task_header(config, &display_opts);
-  display_task_inline(config, uid, &task, true, &display_opts);
```
Comment (phaazon) [inline src/cli.rs:179]: Ditto.

File: src/cli.rs [hunk 4]
```diff
@@ -397,44 +394,33 @@ fn display_task_header(config: &Config, opts: &DisplayOptions) {
...
   let task_status = task.status();
 
   match task_status {
     Status::Todo => {
-      if parity {
-        name = task.name().bright_white().on_black();
-      } else {
-        name = task.name().bright_white().on_bright_black();
-      }
-      status = config.todo_alias().clone().bold().magenta();
```
Comment (phaazon) [inline src/cli.rs:417]: So this is going to be visual regression, as we’re not going to have the alternating color anymore. I Ideally, users should have the possibility to define an alternative color — this could be optional, for instance, and in the absence of this color, we would use the same color for both.

File: src/config.rs [hunk 5]
```diff
@@ -189,3 +200,343 @@ impl Config {
...
+  Underline,
+  Reversed,
+  Italic,
+  Blink,
+  Hidden,
+  Strikethrough,
+}
+
+impl Style {
+  /// returns the function used to apply this `Style` to a `ColoredString`
```
Comment (phaazon) [inline src/config.rs:230]: Please use the same style as all the other functions — starting with an uppercase later and ending with a dot.

File: src/config.rs [hunk 6]
```diff
@@ -189,3 +200,343 @@ impl Config {
...
+      Style::Bold => ColoredString::bold,
+      Style::Dimmed => ColoredString::dimmed,
+      Style::Underline => ColoredString::underline,
+      Style::Reversed => ColoredString::reversed,
+      Style::Italic => ColoredString::italic,
+      Style::Blink => ColoredString::blink,
+      Style::Hidden => ColoredString::hidden,
+      Style::Strikethrough => ColoredString::strikethrough,
+    }
+  }
```
Comment (phaazon) [inline src/config.rs:243]: This is indeed a very odd way to do that. We need to change that. Ideally, a `Style` alone shouldn’t be appliable to a `ColoredString`. An abstraction is needed to go from a `&str` or `String` into a _styled string_. `colored` is inside the implementation of that abstraction, but it shouldn’t leak.

File: src/config.rs [hunk 7]
```diff
@@ -189,3 +200,343 @@ impl Config {
...
+  ($name:ident) => {
+    Color(colored!($name))
+  };
+}
+
+macro_rules! colored {
+  ($name:ident) => {
+    colored::Color::$name
+  };
+}
```
Comment (phaazon) [inline src/config.rs:214]: These macros are not really adding any value, just shortcuts.

File: Cargo.toml [hunk 8]
```diff
@@ -23,5 +23,7 @@ env_logger = "0.8.1"
...
 structopt = "0.3.13"
 toml = "0.5.6"
+scarlet = "1.1.0"
```
Comment (phaazon) [inline Cargo.toml:29]: Please move this before `serde` so that it’s correctly sorted.

File: rfcs/0001-initial-design.md [hunk 9]
```diff
@@ -425,6 +425,59 @@ It contains the following keys:
 - `editor`: if `$EDITOR` is not set, this variable will be used to edit notes. If this variable is set while `$EDITOR`
   is set too, `$EDITOR` has predominance.
 
+## Color Config
```
Comment (phaazon) [inline rfcs/0001-initial-design.md:428]: This belongs to the _Configuration_ section, so you can nest it with `###`.

File: src/config.rs [hunk 10]
```diff
@@ -189,3 +200,343 @@ impl Config {
...
+    if let Some(foreground) = &self.foreground {
+      colored = colored.color(foreground.0);
+    }
+
+    if let Some(background) = &self.background {
+      colored = colored.on_color(background.0);
+    }
+
+    for f in self.styles.iter().map(|s| s.apply_func()) {
+      colored = f(colored);
```
Comment (phaazon) [inline src/config.rs:380]: Here, instead of calling `f`, you can remove the map and simply `match style` and call the right function. Much easier and prevent moving function pointers around.

File: src/config.rs [hunk 11]
```diff
@@ -189,3 +200,343 @@ impl Config {
...
+      colored!(Cyan) => "cyan",
+      colored!(White) => "white",
+      colored!(BrightBlack) => "bright black",
+      colored!(BrightRed) => "bright red",
+      colored!(BrightGreen) => "bright green",
+      colored!(BrightYellow) => "bright yellow",
+      colored!(BrightBlue) => "bright blue",
+      colored!(BrightMagenta) => "bright magenta",
+      colored!(BrightCyan) => "bright cyan",
+      colored!(BrightWhite) => "bright white",
```
Comment (phaazon) [inline src/config.rs:473]: Okay, now I see why you created that macro, but your editor should help you with that :slightly_smiling_face:. You can simply `use colored::Color as WhateverYouLike`.

**Review by phaazon — CHANGES_REQUESTED** [review]
Thank you for this nice contribution! That’s good work! A bunch of things that need to be changed, but most of them should be dealt with quickly as the rest is okay to me. The main issue right now is the lack of abstraction (you handle a dependency’s type around, `ColoredString`, which we cannot merge because it will prevent minor change releases and tie the code too much to the dependency — harder to remove / switch to another crate. Also, returning a `fn` is not needed, so you can just move the `match` block at the current call site of the returned function and it should be all good.

For the rest, there’s a bunch of stuff I haven’t annotated because it’s just rephrasing that I will do before merging so that you can focus on important stuff.

File: src/cli.rs (hunk 2 above)
Comment (LordMZTE) [reply src/cli.rs:138]: we agreed in discord to remove this code temporarily unless i missunderstood something

File: src/cli.rs (hunk 3 above)
Comment (LordMZTE) [reply src/cli.rs:179]: ?

File: src/cli.rs (hunk 4 above)
Comment (LordMZTE) [reply src/cli.rs:417]: this is supposed to be temporary. should i implement that in this PR or later?

File: src/config.rs (hunk 7 above)
Comment (LordMZTE) [reply src/config.rs:214]: yes thats true, but `color!(White)` is much easier to write then `Color(colored::Color::White)`

File: src/config.rs (hunk 5 above)
Comment (LordMZTE) [reply src/config.rs:230]: i'm not sure what you mean. what should i name differently?

File: src/config.rs (hunk 7 above)
Comment (LordMZTE) [reply src/config.rs]: i removed the colored!() macro, but kept the color!() macro, as that one is actually pretty helpful

**Comment by phaazon** [comment]
I just merged your commits from command line by adding a bunch of stuff / resolving the rest of the problems related to merge conflicts. Thanks again! :)
